### PR TITLE
chainHead/follow: Provide multiple block hashes to the initialized event

### PR DESCRIPTION
**Pull request by lexnv**
This PR extends the Initialized event of the chainHead_follow subscription.

Now, the event provides multiple finalized block hashes. This information allows clients that are disconnected, and that want to reconnect, to not lose information about the state of the chain.

At the moment, the spec encourages servers to provide at least 1 minute of finalized blocks (~10 blocks). The users are responsible for unpinning these blocks at a later time. This PR tries to report at least 1 finalized block and at most 16 blocks, if they are available.

Closes: https://github.com/paritytech/polkadot-sdk/issues/3432
cc @paritytech/subxt-team

### CODE REVIEW

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 1]
```diff
@@ -178,13 +187,14 @@ where
...
 	fn get_init_blocks_with_forks(
 		&self,
-		startup_point: &StartupPoint<Block>,
+		finalized: Block::Hash,
 	) -> Result<InitialBlocks<Block>, SubscriptionManagementError> {
 		let blockchain = self.backend.blockchain();
 		let leaves = blockchain.leaves()?;
-		let finalized = startup_point.finalized_hash;
 		let mut pruned_forks = HashSet::new();
 		let mut finalized_block_descendants = Vec::new();
```
[the remark this answers is not in the document]
Comment (lexnv) [reply substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:199]: Have create this PR to handle that edge-case: https://github.com/paritytech/polkadot-sdk/pull/3562 🙏

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 2]
```diff
@@ -42,7 +42,14 @@ use sp_blockchain::{
...
 use sp_runtime::traits::{Block as BlockT, Header as HeaderT, NumberFor};
-use std::{collections::HashSet, sync::Arc};
+use std::{
+	collections::{HashSet, VecDeque},
+	sync::Arc,
+};
+
+/// The maximum number of finalized blocks provided by the
+/// `Initialized` event.
+const MAX_FINALIZDED_BLOCKS: usize = 16;
```
Comment (niklasad1) [inline substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:52]:
```suggestion
const MAX_FINALIZED_BLOCKS: usize = 16;
```

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 3]
```diff
@@ -202,13 +212,40 @@ where
 
 				for pair in blocks.zip(parents) {
```
Comment (niklasad1) [inline substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:213]: Unrelated to this PR but I don't understand the logic with `zip` here.

Parents = blocks.len() + 1

Then the `zip` will ignore the last item in blocks is that indented?

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 4]
```diff
@@ -202,13 +212,40 @@ where
...
 		}
 
-		Ok(InitialBlocks { finalized_block_descendants, pruned_forks })
+		let mut current_block = finalized;
+		// The header of the finalized block must not be pruned.
+		let Some(header) = blockchain.header(current_block)? else {
+			return Err(SubscriptionManagementError::BlockHeaderAbsent);
+		};
+
+		// Report at most `MAX_FINALIZDED_BLOCKS`. Note: The node might not have that many blocks.
```
Comment (niklasad1) [inline substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:229]:
```suggestion
		// Report at most `MAX_FINALIZED_BLOCKS`. Note: The node might not have that many blocks.
```

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 5]
```diff
@@ -202,13 +212,40 @@ where
...
+		let mut finalized_block_hashes = VecDeque::with_capacity(MAX_FINALIZDED_BLOCKS);
```
Comment (niklasad1) [inline substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:230]:
```suggestion
		let mut finalized_block_hashes = VecDeque::with_capacity(MAX_FINALIZED_BLOCKS);
```

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5427969

File: substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs [hunk 6]
```diff
@@ -202,13 +212,40 @@ where
...
+
+		// Report at most `MAX_FINALIZED_BLOCKS`. Note: The node might not have that many blocks.
+		let mut finalized_block_hashes = VecDeque::with_capacity(MAX_FINALIZED_BLOCKS);
+
+		// Pin the finalized block.
+		self.sub_handle.pin_block(&self.sub_id, current_block)?;
+		finalized_block_hashes.push_front(current_block);
+		current_block = *header.parent_hash();
+
+		for _ in 0..MAX_FINALIZDED_BLOCKS - 1 {
```
Comment (niklasad1) [inline substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs:237]:
```suggestion
		for _ in 0..MAX_FINALIZED_BLOCKS - 1 {
```

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5428587

**Review by niklasad1 — APPROVED** [review]

**Comment by athei** [comment]
Removed from queue because it was stuck on flaky test and would run into timeout.

**Comment by lexnv** [comment]
Thanks @athei, IIUC we'd need https://github.com/paritytech/polkadot-sdk/pull/3595 to merge first here 👍

**Comment by athei** [comment]
I am dubious if this actually fixes the problem. It is more of a hail marry.

**Pull request merged**